src/features/upload-file/components/edit.tsx

```diff
@@ -1,45 +1,59 @@
-import React, { FC, useState } from 'react'
+import React, { FC, useState, useEffect } from 'react'
 import {
   EditPropertyProps, DropZone, FormGroup, Label, DropZoneItem,
 } from 'admin-bro'
+import PropertyCustom from '../property-custom.type'
 
 const Edit: FC<EditPropertyProps> = ({ property, record, onChange }) => {
   const { params } = record
-  const fieldName = property.custom.file
-  const { mimeTypes } = property.custom
-  const { maxSize } = property.custom
+  const { custom } = property as unknown as { custom: PropertyCustom }
 
+  const path = params[custom.filePathProperty]
+  const key = params[custom.keyProperty]
+  const file = params[custom.fileProperty]
+
+  const [originalKey, setOriginalKey] = useState(key)
   const [filesToUpload, setFilesToUpload] = useState<Array<File>>([])
-  const [originalFilename] = useState(params.filename)
+
+  useEffect(() => {
+    // it means means that someone hit save and new file has been uploaded
+    // in this case fliesToUpload should be cleared.
+    // This happens when user turns off redirect after new/edit
+    if (key !== originalKey) {
+      setOriginalKey(key)
+      setFilesToUpload([])
+    }
+  }, [key, originalKey])
 
   const onUpload = (files: Array<File>): void => {
     setFilesToUpload(files)
-
-    const newRecord = { ...record }
-    const [file] = files
-    const { file: oldFile, filename, ...other } = newRecord.params
-    const newParams = file
-      ? { ...other, filename: file.name, [fieldName]: file }
-      : { ...other, ...(originalFilename ? { filename: originalFilename } : {}) }
+    const [fileToUpload] = files
 
     onChange({
-      ...newRecord,
-      params: newParams,
+      ...record,
+      params: {
+        ...params,
+        ...(fileToUpload && { [custom.fileProperty]: fileToUpload }),
+      },
     })
   }
 
+  const handleRemove = () => {
```

Comment:
in 3.1 I added support for handing nulls (the thing we discussed on the opensource day) so now removing files is possible as well